samples/bluetooth/hci_uart_async/src/hci_uart_async.c

```diff
@@ -0,0 +1,371 @@
+{
+	/* { H4_CMD, le_16(HCI_CMD_OP_RESET), len=0 } */
+	const uint8_t h4_cmd_reset[] = {0x01, 0x03, 0x0C, 0x00};
+	const uint32_t reset_pattern = sys_get_be32(h4_cmd_reset);
+	int err;
+	struct net_buf *h2c_cmd_reset;
+	uint32_t shift_register = 0;
+
+	LOG_DBG("Looking for reset pattern");
+
+	while (shift_register != reset_pattern) {
+		uint8_t read_byte;
+
+		uart_h2c_rx(&read_byte, sizeof(uint8_t));
+		LOG_DBG("h2c: 0x%02x", read_byte);
+		shift_register = (shift_register * 0x100) + read_byte;
+	}
+
+	LOG_DBG("Pattern found");
+	h2c_cmd_reset = bt_buf_get_tx(BT_BUF_H4, K_FOREVER, h4_cmd_reset, sizeof(h4_cmd_reset));
+	LOG_DBG("Fowarding reset");
+
+	err = bt_send(h2c_cmd_reset);
+	__ASSERT(!err, "Failed to send reset: %d", err);
+}
+
+static void h2c_h4_transport(void)
+{
+	for (;;) {
+		int err;
+		struct net_buf *buf;
+		uint8_t h4_type;
+		uint8_t hdr_size;
+		uint8_t *hdr_buf;
+		uint16_t payload_size;
+
+
+		LOG_DBG("h2c: listening");
+
+		/* Read H4 type. */
+		err = uart_h2c_rx(&h4_type, sizeof(uint8_t));
+
+		if (err) {
+			return;
+		}
+		LOG_DBG("h2c: h4_type %d", h4_type);
+
+		/* Allocate buf. */
+		buf = bt_buf_get_tx(BT_BUF_H4, K_FOREVER, &h4_type, sizeof(h4_type));
+		LOG_DBG("h2c: buf %p", buf);
+
+		if (!buf) {
+			/* `h4_type` was invalid. */
+			__ASSERT_NO_MSG(hci_hdr_size(h4_type) == 0);
+
+			LOG_WRN("bt_buf_get_tx failed h4_type %d", h4_type);
+			return;
+		}
+
+		/* Read HCI header. */
+		hdr_size = hci_hdr_size(h4_type);
+		hdr_buf = net_buf_add(buf, hdr_size);
+
+		err = uart_h2c_rx(hdr_buf, hdr_size);
+		if (err) {
+			net_buf_unref(buf);
+			return;
+		}
+		LOG_HEXDUMP_DBG(hdr_buf, hdr_size, "h2c: hci hdr");
+
+		/* Read HCI payload. */
+		payload_size = hci_payload_size(hdr_buf, h4_type);
+
+		LOG_DBG("h2c: payload_size %u", payload_size);
+
+		if (payload_size <= net_buf_tailroom(buf)) {
+			uint8_t *payload_dst = net_buf_add(buf, payload_size);
+
+			err = uart_h2c_rx(payload_dst, payload_size);
+			if (err) {
+				net_buf_unref(buf);
+				return;
+			}
+			LOG_HEXDUMP_DBG(payload_dst, payload_size, "h2c: hci payload");
+		} else {
+			uint8_t *discard_dst;
+			uint16_t discard_size;
+
+			/* Discard oversize packet. */
+			LOG_WRN("h2c: Discarding oversize h4_type %d payload_size %d.", h4_type,
+				payload_size);
+
+			net_buf_reset(buf);
+			discard_dst = net_buf_tail(buf);
+			discard_size = net_buf_max_len(buf);
+
+			while (payload_size) {
+				uint16_t read_size = MIN(payload_size, discard_size);
+
+				err = uart_h2c_rx(discard_dst, read_size);
+				if (err) {
+					net_buf_unref(buf);
+					return;
+				}
+
+				payload_size -= read_size;
+			}
+
+			net_buf_unref(buf);
+			buf = NULL;
+		}
+
+		LOG_DBG("h2c: packet done");
+
+		/* Route buf to Controller. */
+		if (buf) {
+			err = bt_send(buf);
+			if (err) {
+				/* This is not a transport error. */
+				LOG_ERR("bt_send err %d", err);
+				net_buf_unref(buf);
+				buf = NULL;
+			}
+		}
+
+		k_yield();
```

Comment:
is this needed? Wouldn't you just block this thread on waiting for receiving next packet?

Reply:
It's there in the original `hci_uart` sample.

https://github.com/zephyrproject-rtos/zephyr/blob/168de28b40a4ed7bcd4da8c59d74e2341904fff0/samples/bluetooth/hci_uart/src/main.c#L264-L267

Presumably it gives fairer scheduling. It may be that there is no [reschedule point] in e.g. `k_poll` if it returns without waiting.

[reschedule point]: https://docs.zephyrproject.org/latest/kernel/services/scheduling/index.html#concepts